Objects/typeobject.c

```diff
@@ -1739,6 +1739,8 @@ _PyType_AllocNoTrack(PyTypeObject *type, Py_ssize_t nitems)
     if (presize) {
         ((PyObject **)alloc)[0] = NULL;
         ((PyObject **)alloc)[1] = NULL;
+    }
+    if (PyType_IS_GC(type)) {
         _PyObject_GC_Link(obj);
     }
```

Comment:
What changed here such that we now call `_PyObject_GC_Link()` under different conditions?

Reply:
Previously, `presize != 0` and `PyType_IS_GC(type)` were equivalent because every GC object had a `PyGC_Head`. In the free-threaded build, GC objects do not have a `PyGC_Head` and only objects with managed dicts have a pre-header.